src/client/clientmap.cpp

```diff
@@ -180,17 +204,420 @@ void ClientMap::getBlocksInViewRange(v3s16 cam_pos_nodes,
 			p_nodes_max.Z / MAP_BLOCKSIZE + 1);
 }
 
+class MapBlockFlags
+{
+public:
+	static constexpr u16 CHUNK_INNER_BITS  = 4; // number of bits to address inside a chunk
+	static constexpr u16 CHUNK_EDGE = 16; // 2 ^ CHUNK_SHIFT
+	static constexpr u16 CHUNK_MASK = 0x000Fu; // CHUNK_EDGE-1
+	static constexpr u16 CHUNK_MASK_INV = ~CHUNK_MASK;
+	static constexpr std::size_t CHUNK_VOLUME = CHUNK_EDGE * CHUNK_EDGE * CHUNK_EDGE; // volume of a chunk
+
+	MapBlockFlags(v3s16 min_pos, v3s16 max_pos)
+			: min_pos(min_pos), volume(getAddress(min_pos, max_pos) + 1)
+	{
+		chunks.resize(volume.X * volume.Y * volume.Z);
+	}
+
+	class Chunk
+	{
+	public:
+		inline u8 &getBits(v3s16 pos)
+		{
+			std::size_t address = getAddress(pos);
+			return bits[address];
+		}
+
+	private:
+		inline std::size_t getAddress(v3s16 pos) {
+			std::size_t address = (pos.X & CHUNK_MASK) + (pos.Y & CHUNK_MASK) * CHUNK_EDGE + (pos.Z & CHUNK_MASK) * (CHUNK_EDGE * CHUNK_EDGE);
+			return address;
+		}
+
+		std::array<u8, CHUNK_VOLUME> bits;
+	};
+
+	Chunk &getChunk(v3s16 pos)
+	{
+		v3s16 delta = getAddress(min_pos, pos);
+		std::size_t address = delta.X + delta.Y * volume.X + delta.Z * volume.X * volume.Y;
+		Chunk *chunk = chunks[address].get();
+		if (!chunk) {
+			chunk = new Chunk();
+			chunks[address].reset(chunk);
+		}
+		return *chunk;
+	}
+
+private:
+	// returns the dimensions of an area in chunks, given minimum and maximum edges
+	static inline v3s16 getAddress(v3s16 base, v3s16 pos)
+	{
+		return v3s16((pos .X - base.X) >> CHUNK_INNER_BITS, (pos.Y - base.Y) >> CHUNK_INNER_BITS, (pos.Z - base.Z) >> CHUNK_INNER_BITS);
+	}
+
+	std::vector<std::unique_ptr<Chunk>> chunks;
```

Comment:
No, at view range of 600 (my target), a sparse array uses around 10-15% of memory a vector would use, because vast majority of chunks would be frustum-culled. But a sidelength of 16 blocks is indeed too large for smaller distances, I'm replacing it with 8 (has just 10% overhead at view range of 100).